File: src/main/java/io/enthusia/express/util/SoundFeedback.java
package io.enthusia.express.util;

import java.util.Locale;
import java.util.Set;
import java.util.logging.Logger;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;

/** Best-effort cosmetic feedback. Invalid sound configuration never interrupts mail transitions. */
public final class SoundFeedback {
  public enum Cue {
    PACKAGE_SEND("package-send", "minecraft:block.note_block.pling"),
    PACKAGE_CLAIM("package-claim", "minecraft:entity.item.pickup"),
    LETTER_SEND("letter-send", "minecraft:item.book.page_turn"),
    LETTER_OPEN("letter-open", "minecraft:item.book.page_turn");

    private final String key;
    private final String defaultSound;

    Cue(String key, String defaultSound) {
      this.key = key;
      this.defaultSound = defaultSound;
    }
  }

  private final JavaPlugin plugin;
  private final Set<String> warned = new java.util.HashSet<>();

  public SoundFeedback(JavaPlugin plugin) {
    this.plugin = plugin;
  }

  public void play(Player player, Cue cue) {
    if (!plugin.getConfig().getBoolean("sounds.enabled", true)) return;
    Settings settings = settings(cue);
    if (settings == null) return;
    try {
      player.playSound(
          player.getLocation(), settings.sound(), (float) settings.volume(), (float) settings.pitch());
    } catch (RuntimeException error) {
      warnOnce(
          "sounds." + cue.key,
          "Could not play configured sound at sounds."
              + cue.key
              + ": "
              + error.getMessage());
    }
  }

  public void validate() {
    if (!plugin.getConfig().getBoolean("sounds.enabled", true)) return;
    for (Cue cue : Cue.values()) validateCue(cue);
  }

  private void validateCue(Cue cue) {
    settings(cue);
  }

  private Settings settings(Cue cue) {
    String path = "sounds." + cue.key;
    FileConfiguration config = plugin.getConfig();
    Object soundValue = config.get(path + ".sound", cue.defaultSound);
    Object volumeValue = config.get(path + ".volume", 1.0);
    Object pitchValue = config.get(path + ".pitch", 1.0);
    String sound =
        soundValue instanceof String value ? value.toLowerCase(Locale.ROOT) : "";
    double volume = volumeValue instanceof Number value ? value.doubleValue() : Double.NaN;
    double pitch = pitchValue instanceof Number value ? value.doubleValue() : Double.NaN;
    if (!validSound(sound)
        || !Double.isFinite(volume)
        || volume < 0.0
        || !Double.isFinite(pitch)
        || pitch < 0.5
        || pitch > 2.0) {
      warnOnce(path, "Invalid sound configuration at " + path + "; feedback disabled for it.");
      return null;
    }
    return new Settings(sound, volume, pitch);
  }

  static boolean validSound(String sound) {
    return sound != null && sound.matches("[a-z0-9._-]+:[a-z0-9/._-]+");
  }

  private void warnOnce(String key, String message) {
    if (warned.add(key)) {
      Logger logger = plugin.getLogger();
      if (logger != null) logger.warning(message);
    }
  }

  private record Settings(String sound, double volume, double pitch) {}
}

File: src/main/kotlin/io/enthusia/express/infrastructure/util/SoundFeedback.kt
// Private callback contracts document thread ownership and recovery; CodeRabbit requires method documentation.
@file:Suppress("CommentOverPrivateFunction")

package io.enthusia.express.infrastructure.util

import java.util.Locale
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin

/** Cosmetic configuration failures never interrupt persisted mail operations. */
class SoundFeedback(private val plugin: JavaPlugin) {
    enum class Cue(val key: String, val defaultSound: String) {
        PACKAGE_SEND("package-send", "minecraft:block.note_block.pling"),
        PACKAGE_CLAIM("package-claim", "minecraft:entity.item.pickup"),
        LETTER_SEND("letter-send", "minecraft:item.book.page_turn"),
        LETTER_OPEN("letter-open", "minecraft:item.book.page_turn")
    }

    private val warned = HashSet<String>()
    private data class Settings(val sound: String, val volume: Float, val pitch: Float)

    /** Play a configured success cue for an online player while isolating invalid sound arguments. */
    fun play(player: Player, cue: Cue) {
        if (!plugin.config.getBoolean("sounds.enabled", true)) return
        val settings = settings(cue) ?: return
        try {
            player.playSound(player.location, settings.sound, settings.volume, settings.pitch)
        } catch (error: IllegalArgumentException) {
            warnOnce("sounds.${cue.key}", "Could not play configured sound at sounds.${cue.key}: ${error.message}")
        }
    }

    /** Validate every cue at startup without disabling unrelated valid sounds. */
    fun validate() {
        if (plugin.config.getBoolean("sounds.enabled", true)) Cue.entries.forEach { settings(it) }
    }

    /** Load validated sound, volume and pitch values, returning null for disabled or invalid cues. */
    private fun settings(cue: Cue): Settings? {
        val path = "sounds.${cue.key}"
        val config = plugin.config
        val sound = (config.get("$path.sound", cue.defaultSound) as? String)?.lowercase(Locale.ROOT) ?: ""
        val volume = (config.get("$path.volume", 1.0) as? Number)?.toDouble() ?: Double.NaN
        val pitch = (config.get("$path.pitch", 1.0) as? Number)?.toDouble() ?: Double.NaN
        if (!sound.matches(Regex("[a-z0-9._-]+:[a-z0-9/._-]+")) || !validVolume(volume) || !validPitch(pitch)) {
            warnOnce(path, "Invalid sound configuration at $path; feedback disabled for it.")
            return null
        }
        return Settings(sound, volume.toFloat(), pitch.toFloat())
    }

    /** Accept only finite, nonnegative volumes representable as a float. */
    private fun validVolume(value: Double) = value.isFinite() && value in 0.0..Float.MAX_VALUE.toDouble()

    /** Accept only finite pitches within the supported configured range. */
    private fun validPitch(value: Double) = value.isFinite() && value in 0.5..2.0

    /** Emit each configuration diagnostic at most once. */
    private fun warnOnce(key: String, message: String) {
        if (warned.add(key)) plugin.logger?.warning(message)
    }
}
